lib/service/storage/elasticsearch.js

```diff
@@ -822,6 +822,79 @@ class ElasticSearch extends Service {
     }
   }
 
+  /**
+   * Delete fields of a document and replace it
+   *
+   * @param {String} index - Index name
+   * @param {String} collection - Collection name
+   * @param {String} id - Document id
+   * @param {Array}  fields - Document fields
+   * @param {Object} options - refresh (undefined), userId (null)
+   *
+   * @returns {Promise.<{ _id, _version, _source }>}
+   */
+  deleteFields (
+    index,
+    collection,
+    id,
+    fields,
+    { refresh, userId=null } = {})
+  {
+    const
+      esIndex = this._getESIndex(index, collection),
+      esRequest = {
+        id,
+        index: esIndex,
+      };
+
+    // metadata
+    const _kuzzle_info = {
+      author: getUserId(userId),
+      updatedAt: Date.now(),
+      updater: getUserId(userId)
+    };
+
+    return this._client.exists({ id, index: esIndex })
+      .then(({ body: exists }) => {
+        if (!exists) {
+          throw kerror.get('not_found', id, index, collection);
+        }
+
+        debug('DeleteFields document: %o', esRequest);
+        return this._client.get(esRequest)
+          .then(({ body }) => {
+            for (let field of fields) {
+              _.unset(body._source, field);
+            }
+
+            _.set(body, '_source._kuzzle_info',
```

Comment:
If i'm not wrong, we want not to use lodash. This could be done anotherway without mutating the body object

Reply:
Is there a reason to avoid using lodash ?

Reply:
After researching a bit more I saw a lot of script made to remove a field in a document and they are removing the field straight from the _source of the document

This comes from the ES doc (https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html)
```
POST test/_update/1
{
  "script" : "ctx._source.remove('new_field')"
}
```

So I was wondering, is it really a bad behaviour to mutate the body object ? If so, could you explain me why ?

Reply:
Yes, lodash is slow and the behavior is not explicit.

Reply:
> Is there a reason to avoid using lodash ?

Apart from trying to stay appart from external lib as much as we can and so reduce the build size their is (from my part) not much reasons 😄 

> So I was wondering, is it really a bad behaviour to mutate the body object ? If so, could you explain why ?

From my experience of clean code mutation should be avoided.

- Mutation may lead to unexpected and hard-to-debug issues, where data becomes incorrect somewhere, and you have no idea where it happens.
- Mutation makes code harder to understand: at any time, an array or object may have a different value, so we need to be very careful when reading the code.
- Mutation of function arguments makes the behavior of a function surprising.

Reply:
Ok I understand now why mutations may not be desired, but in that case it seems like there is no other choice (or I might have missed it), since the only way I saw to remove a field is through the update API of ES with a script, but this adds a lot of complexity to generate a working script to remove the field that the user want to be removed, and I dont think this is the way to go, and I dont think we want the user to pass a script to update the document since we couldn't control what he could do with it.